dttools/src/jx.c

```diff
@@ -0,0 +1,425 @@
+	va_start(va, fmt);
+	buffer_putvfstring(B, fmt, va);
+	va_end(va);
+	buffer_dup(B, &str);
+	buffer_free(B);
+
+	j = jx_create(JX_STRING);
+	j->string_value = str;
+
+	return j;
+}
+
+struct jx * jx_integer( jx_int_t integer_value )
+{
+	struct jx *j = jx_create(JX_INTEGER);
+	j->integer_value = integer_value;
+	return j;
+}
+
+struct jx * jx_double( double double_value )
+{
+	struct jx *j = jx_create(JX_DOUBLE);
+	j->double_value = double_value;
+	return j;
+}
+
+struct jx * jx_boolean( int boolean_value )
+{
+	struct jx *j = jx_create(JX_BOOLEAN);
+	j->boolean_value = boolean_value;
+	return j;
+}
+
+struct jx * jx_object( struct jx_pair *pairs )
+{
+	struct jx *j = jx_create(JX_OBJECT);
+	j->pairs = pairs;
+	return j;
+}
+
+struct jx * jx_array( struct jx_item *items )
+{
+	struct jx *j = jx_create(JX_ARRAY);
+	j->items = items;
+	return j;
+}
+
+struct jx * jx_arrayv( struct jx *value, ... )
+{
+	va_list args;
+	va_start(args,value);
+
+	struct jx *array = jx_array(0);
+	while(value) {
+		jx_array_append(array,value);
+		value = va_arg(args,struct jx *);
+	}
+
+	return array;
+}
+
+struct jx * jx_lookup( struct jx *j, const char *key )
+{
+	struct jx_pair *p;
+
+	if(!j || j->type!=JX_OBJECT) return 0;
+
+	for(p=j->pairs;p;p=p->next) {
+		if(p && p->key && p->key->type==JX_STRING) {
+			if(!strcmp(p->key->string_value,key)) {
+				return p->value;
+			}
+		}
+	}
+
+	return 0;
+}
+
+const char * jx_lookup_string( struct jx *object, const char *key )
+{
+	struct jx *j = jx_lookup(object,key);
+	if(j && jx_istype(j,JX_STRING)) {
+		return j->string_value;
+	} else {
+		return 0;
+	}
+}
+
+jx_int_t jx_lookup_integer( struct jx *object, const char *key )
+{
+	struct jx *j = jx_lookup(object,key);
+	if(j && jx_istype(j,JX_INTEGER)) {
+		return j->integer_value;
+	} else {
+		return 0;
+	}
+}
+
+double jx_lookup_double( struct jx *object, const char *key )
+{
+	struct jx *j = jx_lookup(object,key);
+	if(j && jx_istype(j,JX_DOUBLE)) {
+		return j->double_value;
+	} else {
+		return 0;
+	}
+}
+
+int jx_insert( struct jx *j, struct jx *key, struct jx *value )
+{
+	if(!j || j->type!=JX_OBJECT) return 0;
+	j->pairs = jx_pair(key,value,j->pairs);
+	return 1;
+}
+
+void jx_insert_integer( struct jx *j, const char *key, jx_int_t value )
+{
+	jx_insert(j,jx_string(key),jx_integer(value));
+}
+
+void jx_insert_double( struct jx *j, const char *key, double value )
+{
+	jx_insert(j,jx_string(key),jx_double(value));
+}
+
+void jx_insert_string( struct jx *j, const char *key, const char *value )
+{
+	jx_insert(j,jx_string(key),jx_string(value));
+}
+
+void jx_array_insert( struct jx *array, struct jx *value )
+{
+	array->items = jx_item( value, array->items->next );
+}
+
+void jx_array_append( struct jx *array, struct jx *value )
+{
+	if(!array->items) {
+		array->items = jx_item( value, 0 );
+	} else {
+		struct jx_item *i;
+		for(i=array->items;i->next;i=i->next) { }
+		i->next = jx_item(value,0);
+	}
+}
+
+void jx_pair_delete( struct jx_pair *pair )
```

Comment:
Should this and `jx_item` be static/private? I don't think there is a sensible external use.